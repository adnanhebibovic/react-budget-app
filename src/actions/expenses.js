import uuid from 'uuid'

export const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date().getTime()
        }
    }
}

export const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

export const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}