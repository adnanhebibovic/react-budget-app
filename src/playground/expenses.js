import { createStore, combineReducers } from 'redux'
import uuid from 'uuid'

const addExpense = function({title = '', amount = 0} = {}) {
    return {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            title,
            amount,
            createdAt: new Date()
        }
    }
}

const removeExpense = function(id = uuid()) {
    return {
        type: 'REMOVE_EXPENSE',
        id
    }
}

const editExpense = function(id = uuid(), expense = {}) {
    return {
        type: 'EDIT_EXPENSE',
        id,
        expense
    }
}

const expensesDefaultState = []

const expensesReducer = function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}

const setTextFilter = function(text = undefined) {
    return {
        type: 'SET_TEXT_FILTER',
        text
    }
}

const sortByAmount = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'amount'
    }
}

const sortByDate = function() {
    return {
        type: 'SORT_BY',
        sortBy: 'createdAt'
    }
}

const setDateFilters = function({startDate = undefined, endDate = undefined} = {}) {
    return {
        type: 'SET_DATE_FILTERS',
        dates: {
            startDate,
            endDate
        }
    }
}

const filtersDefaultState = {
    text: undefined,
    sortBy: undefined,
    startDate: undefined,
    endDate: undefined
}

const filtersReducer = function(state = filtersDefaultState, action) {
    switch (action.type) {
        case 'SET_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            }
        case 'SORT_BY':
            return {
                ...state,
                sortBy: action.sortBy
            }
        case 'SET_DATE_FILTERS':
            return {
                ...state,
                ...action.dates
            }
        default:
            return state;
    }
}

const getExpenses = function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text === 'undefined' || expense.title.includes(text);
        const startDateMatch = typeof startDate === 'undefined' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate === 'undefined' || expense.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy)
    {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result
}

const store = createStore(
    combineReducers({
        expenses: expensesReducer,
        filters: filtersReducer
    })
)

store.dispatch(addExpense({
    title: 'January bills',
    amount: 279.00 
}))

store.dispatch(addExpense({
    title: 'February bills',
    amount: 523.00 
}))

store.dispatch(addExpense({
    title: 'March bills',
    amount: 358.00 
}))

store.dispatch(addExpense({
    title: 'New car',
    amount: 1000.00 
}))

store.subscribe(() => {
    const state = store.getState();
    console.log(getExpenses(state.expenses, state.filters))
})

store.dispatch(setTextFilter('bills'))
store.dispatch(sortByAmount())
store.dispatch(sortByDate())
store.dispatch(setDateFilters({endDate: new Date()}))
