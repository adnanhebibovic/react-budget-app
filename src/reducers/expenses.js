export const expensesDefaultState = []

export default function(state = expensesDefaultState, action) {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return state.concat(action.expense)
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => expense.id == action.id 
                ? {...expense, ...action.expense}
                : expense
            )
        default:
            return state;
    }
}