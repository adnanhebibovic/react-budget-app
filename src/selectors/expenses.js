export default function(expenses, {text, sortBy, startDate, endDate}) {
    const result = expenses.filter((expense) => {
        const textMatch = typeof text == 'undefined' || expense.title.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = typeof startDate == 'undefined' || expense.createdAt > startDate;
        const endDateMatch = typeof endDate == 'undefined' || expense.createdAt < endDate;

        return textMatch && startDateMatch && endDateMatch;
    })

    if (sortBy) {
        return result.sort((e1, e2) => {
            if (e1[sortBy] < e2[sortBy]) {
                return -1;
            } else if (e1[sortBy] > e2[sortBy]) {
                return 1;
            } else {
                return 0;
            }
        })
    }

    return result;
}
