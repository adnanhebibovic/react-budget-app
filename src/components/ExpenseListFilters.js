import React, { useState } from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount, setDateFilters } from '../actions/filters'

function ExpenseListFilters(props) {
    function onTextChange(e) {
        props.setTextFilter(e.target.value)
    }

    function onSelectChange(e) {
        switch(e.target.value) {
            case 'createdAt':
                props.sortByDate()
                return;
            case 'amount':
                props.sortByAmount()
                return;
        }
    }

    function onStartDateChange(e) {
        const startDate = new Date(e.target.value)

        setStartDate(startDate)
        props.setDateFilters(startDate, endDate)
    }

    function onEndDateChange(e) {
        const endDate = new Date(e.target.value)

        setEndDate(endDate)
        props.setDateFilters(startDate, endDate)
    }

    const [startDate, setStartDate] = useState(undefined)
    const [endDate, setEndDate] = useState(undefined)

    return (
        <div>
            <input type="text" defaultValue={ props.filters.text } onChange={onTextChange}></input>
            
            <select defaultValue={ props.filters.sortBy } onChange={onSelectChange}>
                <option value="createdAt">Date</option>
                <option value="amount">Amount</option>
            </select>

            <input type="date" name="startDate" onChange={onStartDateChange}></input>
            <input type="date" name="endDate" onChange={onEndDateChange}></input>
        </div>
    )
}

function mapStateToProps(state) {
    return {
        filters: state.filters
    }
}

function mapDispatchToProps(dispatch) {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByAmount: () => dispatch(sortByAmount()),
        sortByDate: () => dispatch(sortByDate()),
        setDateFilters: (startDate, endDate) => dispatch(setDateFilters(startDate, endDate))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);