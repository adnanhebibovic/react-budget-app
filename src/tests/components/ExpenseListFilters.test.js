import React from 'react';
import { Provider } from 'react-redux';
import { create, act } from 'react-test-renderer';
import configureStore from '../../store/expense-store'

import expenses from '../fixtures/expenses';

import ExpenseListFilters from '../../components/ExpenseListFilters';

let store;
let renderer;

beforeEach(() => {
    const filters = {
        text: undefined,
        sortBy: 'createdAt',
        startDate: undefined,
        endDate: undefined
    }

    store = configureStore(expenses, filters);

    renderer = create(
        <Provider store={store}>
            <ExpenseListFilters />
        </Provider>
    );
});  

test('Should render ExpenseListFilters component correctly', () => {
    expect(renderer).toMatchSnapshot();
})

test('Should dispatch a setText action filter with changed data', () => {
    const filters = {
        text: expenses[0].title,
        sortBy: 'createdAt',
        startDate: undefined,
        endDate: undefined
    }

    const instance = renderer.root;

    act(() => {
        instance.findByProps({type: 'text'}).props.onChange({ target: { value: expenses[0].title } });
    });

    let state = store.getState().filters
    
    expect(state).toEqual(filters)
})

test('Should dispatch a sortByAmount action filter with changed data', () => {
    const filters = {
        text: undefined,
        sortBy: 'amount',
        startDate: undefined,
        endDate: undefined
    }

    const instance = renderer.root;

    act(() => {
        instance.findByType('select').props.onChange({ target: { value: 'amount' } });
    });

    let state = store.getState().filters
    
    expect(state).toEqual(filters)
})

test('Should dispatch a setDateFilters action filter with changed data', () => {
    let filters = {
        text: undefined,
        sortBy: 'createdAt',
        startDate: new Date('2021-01-01'),
        endDate: undefined
    }

    const instance = renderer.root;

    act(() => {
        instance.findByProps({name: 'startDate'}).props.onChange({ target: { value: new Date('2021-01-01') } });
    });

    let state = store.getState().filters
    
    expect(state).toEqual(filters)

    filters = {
        text: undefined,
        sortBy: 'createdAt',
        startDate: new Date('2021-01-01'),
        endDate: new Date('2021-12-31')
    }

    act(() => {
        instance.findByProps({name: 'endDate'}).props.onChange({ target: { value: new Date('2021-12-31') } });
    });

    state = store.getState().filters

    expect(state).toEqual(filters)
})