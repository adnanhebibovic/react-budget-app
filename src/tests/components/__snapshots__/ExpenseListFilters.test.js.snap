// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListFilters component correctly 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
  />
  <select
    defaultValue="createdAt"
    onChange={[Function]}
  >
    <option
      value="createdAt"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <input
    name="startDate"
    onChange={[Function]}
    type="date"
  />
  <input
    name="endDate"
    onChange={[Function]}
    type="date"
  />
</div>
`;
